src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -0,0 +1,506 @@
+                'tax_excl' => $rule->getContextualValue(false),
+            );
+            $orderCartRule = new OrderCartRule();
+            $orderCartRule->id_order = $order->id;
+            $orderCartRule->id_cart_rule = $cartRule['id_cart_rule'];
+            $orderCartRule->id_order_invoice = $invoice->id;
+            $orderCartRule->name = $cartRule['name'];
+            $orderCartRule->value = $values['tax_incl'];
+            $orderCartRule->value_tax_excl = $values['tax_excl'];
+            $orderCartRule->add();
+
+            $order->total_discounts += $orderCartRule->value;
+            $order->total_discounts_tax_incl += $orderCartRule->value;
+            $order->total_discounts_tax_excl += $orderCartRule->value_tax_excl;
+            $order->total_paid -= $orderCartRule->value;
+            $order->total_paid_tax_incl -= $orderCartRule->value;
+            $order->total_paid_tax_excl -= $orderCartRule->value_tax_excl;
+        }
+
+        $order->update();
+    }
+
+    /**
+     * @param Order $order
+     *
+     * @throws OrderException
+     */
+    private function assertOrderWasNotShipped(Order $order)
+    {
+        if ($order->hasBeenShipped()) {
+            throw new OrderException('Cannot add product to shipped order.');
+        }
+    }
+
+    /**
+     * @param int $combinationId
+     *
+     * @return Combination|null
+     */
+    private function getCombination($combinationId)
+    {
+        $combination = null;
+
+        if (0 !== $combinationId) {
+            $combination = new Combination($combinationId);
+
+            if (Validate::isLoadedObject($combination)) {
+                throw new OrderException('Product combination not found.');
+            }
+        }
+
+        return $combination;
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param int $langId
+     *
+     * @return Product
+     */
+    private function getProductObject(ProductId $productId, $langId)
+    {
+        $product = new Product($productId->getValue(), false, $langId);
+
+        if ($product->id !== $productId->getValue()) {
+            throw new OrderException(
+                sprintf('Product with id "%d" is invalid.', $productId->getValue())
+            );
+        }
+
+        return $product;
+    }
+
+    /**
+     * @param Order $order
+     *
+     * @return Cart
+     */
+    private function createNewCart(Order $order)
+    {
+        $cart = new Cart();
+        $cart->id_shop_group = $order->id_shop_group;
+        $cart->id_shop = $order->id_shop;
+        $cart->id_customer = $order->id_customer;
+        $cart->id_carrier = $order->id_carrier;
+        $cart->id_address_delivery = $order->id_address_delivery;
+        $cart->id_address_invoice = $order->id_address_invoice;
+        $cart->id_currency = $order->id_currency;
+        $cart->id_lang = $order->id_lang;
+        $cart->secure_key = $order->secure_key;
+
+        $cart->add();
+
+        Context::getContext()->cart = $cart;
```

Comment:
Created https://github.com/PrestaShop/PrestaShop/issues/14131 to solve this in the future